File: src/references/referenceParser.js
// @ts-check

import { REFERENCE_TYPE } from "./referenceTypes.js";

/**
 * Documentation reference syntax recognized inside comments:
 *
 *   documentation/file.md
 *   documentation/file.md#anchor
 *   documentation/file.md - anchor
 *   documentation/file.md — anchor
 *   documentation/file.md:42
 *   documentation/file.md#L42
 *   documentation/file.md#l42
 *
 * `file.md#42` is deliberately NOT treated as a line reference — it stays a
 * heading anchor so `#anchor` and `#L42` never conflict.
 *
 * The path is not allowed to start with `/` (absolute paths are rejected) and
 * the look-behind prevents false positives inside URLs
 * (`https://host/docs/file.md`), inside longer identifiers and after a
 * backslash (`C:\docs\file.md`, UNC paths).
 */
const DOCUMENTATION_REFERENCE_REGEX = new RegExp(
  `(?<![\\w:./\\\\])([A-Za-z0-9_.-][A-Za-z0-9_./-]*\\.md)` +
    `(?:(?::(\\d+))|` +
    `(?:#[Ll](\\d+))|` +
    `(?:#|\\s+-\\s+|\\s+\u2014\\s+)([A-Za-z0-9_-]+))?`,
  "g",
);

/**
 * Anchored variants used by {@link parseReference} to normalize a single
 * already-detected reference. Each group order mirrors the detection regex.
 */
const DOCUMENTATION_REFERENCE_ANCHORED = new RegExp(
  `^([A-Za-z0-9_.-][A-Za-z0-9_./-]*\\.md)` +
    `(?:(?::(\\d+))|` +
    `(?:#[Ll](\\d+))|` +
    `(?:#|\\s+-\\s+|\\s+\u2014\\s+)([A-Za-z0-9_-]+))?$`,
);

/**
 * Issue reference: `#123`.
 *
 * A leading word character is forbidden so `foo#123` is not an issue
 * reference and `file.md#L42` is not partially re-detected.
 */
const ISSUE_REFERENCE_REGEX = /(?<![\w:#])#(\d+)\b/g;

const ISSUE_REFERENCE_ANCHORED = /^#(\d+)$/;

/**
 * API reference: `API:Foo`.
 */
const API_REFERENCE_REGEX = /(?<!\w)API:([A-Za-z0-9_-]+)\b/g;

const API_REFERENCE_ANCHORED = /^API:([A-Za-z0-9_-]+)$/;

/**
 * @typedef {object} TicketLink
 * @property {string} baseUrl URL prefix appended with the matched key.
 * @property {RegExp} regex Compiled ticket-key pattern.
 * @property {string|null} label Optional hover label.
 */

/**
 * @typedef {object} ReferenceSpan
 * @property {string} raw
 * @property {number} start
 * @property {number} end
 */

/**
 * @typedef {object} ReferenceSpan
 * @property {string} raw
 * @property {number} start
 * @property {number} end
 * @property {string} [url] Resolved click URL for ticket references.
 * @property {string|null} [label] Hover label for ticket references.
 */

/**
 * Detect reference spans in a piece of text, in priority order.
 *
 * Documentation references win over the generic issue/ticket/API forms so a
 * reference inside an already-matched span (for example `file.md#123`) is
 * never reported twice with conflicting types. Ticket keys from
 * `ticketLinks` are detected last; a span already consumed by a higher-priority
 * reference is skipped, and the first matching ticket entry wins.
 *
 * @param {string} text
 * @param {TicketLink[]} [ticketLinks]
 * @returns {ReferenceSpan[]} Spans sorted by start offset.
 */
export function detectReferenceSpans(text, ticketLinks = []) {
  const accepted = [];
  const consumed = [];

  const accept = (match, meta) => {
    const span = {
      raw: match[0],
      start: match.index,
      end: match.index + match[0].length,
    };

    if (meta) {
      span.url = meta.url;
      span.label = meta.label;
    }

    for (const existing of consumed) {
      if (span.start < existing.end && existing.start < span.end) {
        return;
      }
    }

    consumed.push(span);
    accepted.push(span);
  };

  for (const match of text.matchAll(DOCUMENTATION_REFERENCE_REGEX)) {
    accept(match);
  }

  for (const match of text.matchAll(ISSUE_REFERENCE_REGEX)) {
    accept(match);
  }

  for (const match of text.matchAll(API_REFERENCE_REGEX)) {
    accept(match);
  }

  for (const entry of ticketLinks) {
    for (const match of text.matchAll(entry.regex)) {
      accept(match, { url: entry.baseUrl + match[0], label: entry.label });
    }
  }

  return accepted.sort((a, b) => a.start - b.start);
}

/**
 * Normalize a raw reference string into a typed reference.
 *
 * A documentation reference carries a `file` plus optional `anchor`/`line`.
 * Issue/API/DOC-… references carry an `identifier`.
 *
 * @param {string} raw
 * @returns {{
 *   type: string,
 *   raw: string,
 *   file: string|null,
 *   anchor: string|null,
 *   line: number|null,
 *   identifier: string|null,
 *   url: string|null,
 *   label: string|null
 * }|null}
 */
export function parseReference(raw) {
  const documentation = raw.match(DOCUMENTATION_REFERENCE_ANCHORED);

  if (documentation) {
    const line =
      documentation[2] !== undefined
        ? Number(documentation[2])
        : documentation[3] !== undefined
          ? Number(documentation[3])
          : null;

    return {
      type: REFERENCE_TYPE.DOCUMENTATION,
      raw,
      file: documentation[1],
      anchor: documentation[4] ?? null,
      line,
      identifier: null,
    };
  }

  const issue = raw.match(ISSUE_REFERENCE_ANCHORED);

  if (issue) {
    return {
      type: REFERENCE_TYPE.ISSUE,
      raw,
      file: null,
      anchor: null,
      line: null,
      identifier: issue[1],
    };
  }

  const api = raw.match(API_REFERENCE_ANCHORED);

  if (api) {
    return {
      type: REFERENCE_TYPE.API,
      raw,
      file: null,
      anchor: null,
      line: null,
      identifier: api[1],
    };
  }

  return null;
}

/**
 * @typedef {object} ParsedReference
 * @property {string} type
 * @property {string} raw
 * @property {string|null} file
 * @property {string|null} anchor
 * @property {number|null} line
 * @property {string|null} identifier
 * @property {string|null} url Resolved click URL (ticket references only).
 * @property {string|null} label Hover label (ticket references only).
 * @property {number} start
 * @property {number} end
 */

/**
 * Parse every reference found in a comment text.
 *
 * Ticket references are produced directly from the spans detected by
 * {@link detectReferenceSpans} (which carries the resolved URL/label), so they
 * never rely on `parseReference`. `parseReference` handles documentation,
 * issue and API references only.
 *
 * Offsets are relative to `offset`, which should be the position of the
 * comment text inside its containing line.
 *
 * @param {string} text
 * @param {number} [offset]
 * @param {TicketLink[]} [ticketLinks]
 * @returns {ParsedReference[]}
 */
export function parseComment(text, offset = 0, ticketLinks = []) {
  const references = [];

  for (const span of detectReferenceSpans(text, ticketLinks)) {
    /** @type {object} */
    let parsed;

    if (span.url !== undefined && span.url !== null) {
      parsed = {
        type: REFERENCE_TYPE.TICKET,
        raw: span.raw,
        file: null,
        anchor: null,
        line: null,
        identifier: span.raw,
        url: span.url,
        label: span.label ?? null,
      };
    } else {
      parsed = parseReference(span.raw);

      if (parsed === null) {
        continue;
      }
    }

    references.push({
      ...parsed,
      start: offset + span.start,
      end: offset + span.end,
    });
  }

  return references;
}

File: jetbrains/src/main/kotlin/com/anchor/commentdoclinks/parser/ReferenceParser.kt
package com.anchor.commentdoclinks.parser

import com.anchor.commentdoclinks.model.ParsedReference
import com.anchor.commentdoclinks.model.ReferenceType
import com.anchor.commentdoclinks.model.TicketLink

/**
 * Documentation reference detection (global).
 *
 * The path must not start with `/` (absolute paths rejected) and the
 * look-behind rejects references inside URLs, after word chars, `:`, `.`,
 * `/`, and `\` (so `C:\docs\file.md` and `\\server\share\docs\file.md` are
 * rejected). `#42` is deliberately NOT a line reference — it stays a heading
 * anchor so `#anchor` and `#L42` never conflict.
 */
private val DOCUMENTATION_REGEX =
    Regex(
        """(?<![\w:./\\])([A-Za-z0-9_.-][A-Za-z0-9_./-]*\.md)""" +
            """(?:(?::(\d+))|(?:#[Ll](\d+))|(?:#|\s+-\s+|\s+—\s+)([A-Za-z0-9_-]+))?""",
    )

/** Anchored variant used to normalize a single already-detected reference. */
private val DOCUMENTATION_ANCHORED =
    Regex(
        """^([A-Za-z0-9_.-][A-Za-z0-9_./-]*\.md)""" +
            """(?:(?::(\d+))|(?:#[Ll](\d+))|(?:#|\s+-\s+|\s+—\s+)([A-Za-z0-9_-]+))?$""",
    )

/** Issue reference: `#123` (leading word char forbidden). */
private val ISSUE_REGEX = Regex("""(?<![\w:#])#(\d+)\b""")

private val ISSUE_ANCHORED = Regex("""^#(\d+)$""")

/** API reference: `API:Foo`. */
private val API_REGEX = Regex("""(?<!\w)API:([A-Za-z0-9_-]+)\b""")

private val API_ANCHORED = Regex("""^API:([A-Za-z0-9_-]+)$""")

/**
 * A raw reference span with 0-based character offsets.
 *
 * @property url Resolved click URL for ticket references; `null` for all other
 *   reference kinds.
 * @property label Hover label for ticket references; `null` otherwise.
 */
data class ReferenceSpan(
    val raw: String,
    val start: Int,
    val end: Int,
    val url: String? = null,
    val label: String? = null,
)

/**
 * Compile a user-supplied ticket-key pattern into a detection regex.
 *
 * The pattern is wrapped with a leading `(?<!\w)` look-behind so keys embedded
 * in longer words or URLs are not matched. Invalid patterns are skipped
 * (logged) to guard against bad regex / ReDoS from user input.
 */
internal fun TicketLink.toRegex(): Regex? =
    try {
        Regex("""(?<!\w)($pattern)(?!\w)""")
    } catch (e: Exception) {
        println("commentDocLinks.ticketLinks: skipping invalid pattern \"$pattern\": ${e.message}")
        null
    }

/**
 * Detect reference spans in text, in priority order.
 *
 * Documentation references win over the generic issue/API forms, so a
 * reference inside an already-matched span (for example `file.md#123`) is
 * never reported twice with conflicting types. Ticket keys from
 * [ticketLinks] are detected last; a span already consumed by a
 * higher-priority reference is skipped and the first matching ticket entry
 * wins.
 *
 * @return spans sorted by start offset
 */
fun detectReferenceSpans(
    text: String,
    ticketLinks: List<TicketLink> = emptyList(),
): List<ReferenceSpan> {
    val accepted = mutableListOf<ReferenceSpan>()
    val consumed = mutableListOf<ReferenceSpan>()

    fun accept(
        match: MatchResult,
        url: String? = null,
        label: String? = null,
    ) {
        val span =
            ReferenceSpan(
                raw = match.value,
                start = match.range.first,
                end = match.range.last + 1,
                url = url,
                label = label,
            )
        for (existing in consumed) {
            if (span.start < existing.end && existing.start < span.end) return
        }
        consumed.add(span)
        accepted.add(span)
    }

    DOCUMENTATION_REGEX.findAll(text).forEach { match -> accept(match) }
    ISSUE_REGEX.findAll(text).forEach { match -> accept(match) }
    API_REGEX.findAll(text).forEach { match -> accept(match) }

    for (link in ticketLinks) {
        val regex = link.toRegex() ?: continue
        regex.findAll(text).forEach { match ->
            accept(match, link.baseUrl + match.value, link.label)
        }
    }

    return accepted.sortedBy { it.start }
}

/**
 * Normalize a raw reference string into a typed [ParsedReference].
 *
 * Handles documentation, issue and API references. Ticket references are
 * produced directly by [parseComment] from the spans detected by
 * [detectReferenceSpans] (which carry the resolved URL/label), so they never
 * rely on this function.
 */
fun parseReference(raw: String): ParsedReference? {
    DOCUMENTATION_ANCHORED.matchEntire(raw)?.let { m ->
        val g = m.groupValues
        val line =
            when {
                g[2].isNotEmpty() -> g[2].toIntOrNull()
                g[3].isNotEmpty() -> g[3].toIntOrNull()
                else -> null
            }
        return ParsedReference(
            type = ReferenceType.DOCUMENTATION,
            raw = raw,
            file = g[1],
            anchor = if (g[4].isNotEmpty()) g[4] else null,
            line = line,
            identifier = null,
        )
    }

    ISSUE_ANCHORED.matchEntire(raw)?.let { m ->
        return ParsedReference(
            type = ReferenceType.ISSUE,
            raw = raw,
            file = null,
            anchor = null,
            line = null,
            identifier = m.groupValues[1],
        )
    }

    API_ANCHORED.matchEntire(raw)?.let { m ->
        return ParsedReference(
            type = ReferenceType.API,
            raw = raw,
            file = null,
            anchor = null,
            line = null,
            identifier = m.groupValues[1],
        )
    }

    return null
}

/**
 * Parse every reference found in comment text.
 *
 * Ticket references are produced directly from the spans detected by
 * [detectReferenceSpans] (which carries the resolved URL/label).
 *
 * Offsets are relative to [offset], which should be the position of the
 * comment text inside its containing line.
 */
fun parseComment(
    text: String,
    offset: Int = 0,
    ticketLinks: List<TicketLink> = emptyList(),
): List<ParsedReference> =
    detectReferenceSpans(text, ticketLinks).mapNotNull { span ->
        val parsed =
            if (span.url != null) {
                ParsedReference(
                    type = ReferenceType.TICKET,
                    raw = span.raw,
                    file = null,
                    anchor = null,
                    line = null,
                    identifier = span.raw,
                    url = span.url,
                    label = span.label,
                )
            } else {
                parseReference(span.raw) ?: return@mapNotNull null
            }
        parsed.copy(start = offset + span.start, end = offset + span.end)
    }
